### DISCUSSION TITLE
How to access router

### DISCUSSION THREAD

**Discussion by grst** [froonix/HH40V#3] [General]
Hi, 

very interesting analysis! 
How did you manage to access the router? 
I get a connection refused when I try to connect via ethernet. 
```
>ssh 192.168.2.200 
ssh: connect to host 192.168.2.200 port 22: Connection refused
```

(I have a T-Mobile branded version, they are probably trying to keep me out. But maybe you found a way around ;) )

**Comment by killerbees19** [comment]
I've not used the device since mid-2018, maybe they changed something at the firmware. I'm even unsure if it was SSH or Telnet to access the device. Sorry! 😄

You could try to run nmap to find open ports. If you want share your results afterwards. Thanks! 🙂

**Comment by grst** [comment]
But you were connecting via Ethernet cable, right?
Just asking because the device was called `usb1`

**Comment by killerbees19** [comment]
There are to different systems running:

* Connection via ethernet to access the device itself (OpenWrt)
* Connection to the internal modem (Android) – see [network](https://github.com/froonix/HH40V/wiki/network-info) and [RNDIS](https://github.com/froonix/HH40V/wiki/RNDIS-device)

<s>Afaik you could connect the device via µUSB too, without an ethernet connection. Someone told me this detail by mail but I've not tested it yet.</s>

**Comment by killerbees19** [comment]
> But you were connecting via Ethernet cable, right?

To answer your question precisely: Yes! :-)

**Comment by grst** [comment]
here's the result of my nmap. 
The router doesn't react on telnet for all ports except 80 and 7547. 
So probably not direct ssh. Might be possible to get into it via the cwmp. 

```
> nmap -p- 192.168.3.200
Nmap scan report for hh40.home (192.168.3.200)
Host is up (0.0011s latency).
Not shown: 65526 closed ports
PORT     STATE    SERVICE
53/tcp   open     domain
80/tcp   open     http
135/tcp  filtered msrpc
137/tcp  filtered netbios-ns
138/tcp  filtered netbios-dgm
139/tcp  filtered netbios-ssn
445/tcp  filtered microsoft-ds
5555/tcp filtered freeciv
7547/tcp open     cwmp
```

**Comment by grst** [comment]
FWIW, I now just put a FritzBox behind the HH40V. The Magenta-branded firmware just sucks too much.

**Comment by LoboHacks** [comment]
@grst  fyi.. just ssh to 192.168.225.1, works also from wifi.
or add a route to 192.168.225/32 via HH40-IP in your firtzbox.

cheers
oliver

**Comment by LoboHacks** [comment]
seems to be a dead end for ssh access to the router :(

```sh
$ ssh -D 1080 -fNn root@192.168.225.1
$ nmap -p- 192.168.225.111 --proxy socks4://127.0.0.1:1080
Nmap scan report for 192.168.225.111
Host is up (0.0011s latency).
Not shown: 65532 closed ports
PORT     STATE SERVICE
53/tcp   open  domain
80/tcp   open  http
7547/tcp open  cwmp

Nmap done: 1 IP address (1 host up) scanned in 2626.61 seconds
```

**Comment by jamesmacwhite** [comment]
The EE branded HH70 has a button trigger where by if you hold the WPS button for 15 or more seconds, this trigger dropbear to be started. See more info here:

https://github.com/jamesmacwhite/hh70-ee/wiki/SSH-access

I don't know if the HH40 will be the same. Alternatively, there are weaknesses in the backup and restore functionality which you can use to write your own config files to the router filesystem.

**Comment by LoboHacks** [comment]
Unfortunately this doesn't work on the HH40, at least not on mine.
I already tried my luck with the backup file.

binwalk: configure.bin
```bash
DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
30481         0x7711          StuffIt Deluxe Segment (data): f
39020         0x986C          LANCOM OEM file
```
..mööp

binwalk: after base64 -D
```bash
DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             OpenSSL encryption, salted, salt: 0xF1F1E04970265B60
```

..lost

didn't try to brute the encryption yet.

**Comment by jamesmacwhite** [comment]
That's a shame. Looks like the configure.bin has stronger protections than the HH70 I have has. The button trigger was worth a shot, I didn't know it even existed until I had root access back to explore the filesystem. It might be something specific to this variant. I don't know anyone else with HH70/VH.

Looks like your configure.bin might have more checks on it as well, given there's encryption, possible CRC checks as well. The one from the HH70VB has little to no checks, which is odd considering it's from the same company and I'd figure the firmware would be similar.

**Comment by dandare1980** [comment]
I have a BT 4G Home Hub which is a rebranded version of the Alcatel HH70.

The WPS button trick allows me access the OpenWRT filesystem via SSH. To access the modem, I have to login as root to 192.168.1.1 and then SSH again as root to 192.168.225.1.

I would love to know how to connect to this device by USB - when you plug it in to a computer it remains completely undetected.

**Comment by jamesmacwhite** [comment]
@dandare1980 For some reason there are firewall rules to stop direct access to the modem at 192.168.225.1 from the LAN side. It is kind of pointless as you can bypass this by logging into the router first then to the modem as you've found out or alternatively using the IPv4 address configured by the modem which is usually 19.x.x.x. (visible in the web interface).

It may not support direct access for USB or require some debug mode enabling to do it. There is adb, but it seems to lock up when you try and access TCP 5555.

**Comment by dandare1980** [comment]
@jamesmacwhite thank you for pointing me in the direction of the firewall rules.

I decided to comment out the /etc/config/firewall rule which prevented SSH access to the modem located under config rule `option name 'drop-ssh-forward'` with `option dest_ip '192.168.225.1'`.

The file system was for some reason read only so I had to perform `mount -o remount,rw /` first to save the file.

After rebooting the router I could login to the modem part of the router with `ssh root@192.168.225.1` and the usual password.

Would there be any harm in leaving the SSH access to the modem enabled once I've finished what I am doing?

**Comment by jamesmacwhite** [comment]
That will work you could also just add `option enabled '0'` on each rule to save you having to comment out the rule entirely, if you then wanted to toggle it back on quickly, flip it to `1`.

Interesting the BT variant is read only by default, the EE one isn't. I assumed it's literally just the logo and different front and back plate on the BT one that's the major difference, maybe they've been tweaking the firmware as well! I know the flash space is very limited literally has about 100kb left on the rootfs.

Well technically the modem has always had SSH enabled, the firewall rules from the router side will have prevented SSH access on the LAN, however now that you've dropped those rules you'll be able to access it, without having to go through the router first now. Technically it's also been possible to use the 19.x.x.x IPv4 address configured to access the modem as the firewall rules don't protect that firewall zone ¯\_(ツ)_/¯, so it's not exactly been that secure!

Obviously, given the hardcoded credentials, it would be unwise to leave it open on your LAN just hanging out there. For the router side you can at least change the root password or even better configure public key auth without breaking anything. The modem not so much as several binaries have hard coded sshpass calls, which will probably start breaking if you tried to change the oelinux123 password for the root user on the modem side.

I never poked the modem part that much. I was more interested in breaking back into the device to get SSH back, which I did then just poked around a bit after, but I'm sure there's interesting stuff to poke on the modem side for sure. Feel free to contact me privately if you find anything interesting.

**Comment by dandare1980** [comment]
The `option enabled '0'` switch worked well, so for the time being I have changed it to `'1'` to disable SSH access to the modem as it's easy to turn back on again - a great tip thanks.

My router's firmware is listed as `HH70_HR_02.00_02` and when the online update is run from the web interface it reports "Your software is up to date."

Changing the root and admin passwords with `passwd username` was easy enough, but I'll explore the public key auth option another time. I left the modem password well alone as you suggested.

The default web interface doesn't allow you to define your own DNS servers. Can this be done through `/var/run/resolv.conf` on the modem side or are you limited by the network providers selection?

Lastly, is there a config where I can edit the wired ethernet adaptor's name? My computer currently shows itself as connected to "BT4G_HOMEHUB_****_2.4GHz", which is the same default SSID as if it were connected by wireless (I have since changed those defaults).

**Comment by jamesmacwhite** [comment]
@dandare1980 That's a different firmware build for sure. I'd imagine pretty similar to the EE one, but different by the sounds of it. For comparison this is the EE one: HH70_E1_02.00_24, but it's had a lot more revisions given it's the life cycle of the device, although it looks like October 2018 seems to be the date the last major changes were made for the EE builds based on the file system.

You can technically change the DNS resolvers given you've got root access, but this will potentially break things, if you are using IPv6. The DNS servers configured are NAT64 and you'll need to have NAT64 resolvers for domains to resolve properly when they don't have an AAAA record potentially. The EE network is weird on this. For example, it doesn't do graceful fallback for the scenario when a domain might have an AAAA record but it times out. Most browsers would simply go back to IPv4 with the happy eyeballs standard, on EE the site will just fail to load and never fallback to IPv4. You can confirm it by doing a test at ipv6-test.com. The graceful fallback test always fails.

I believe the JSON RPC API may expose a method to set DNS values even though the interface doesn't. I think I remember spotting it, but never tested it. Possibly the SDK call of SetLanSettings is the one. The get equivalent returns these keys, related to DNS, a mode and two keys for storing what looks like resolvers.

```json
{
    "jsonrpc": "2.0",
    "result": {
        "DNSMode": 0,
        "DNSAddress1": "",
        "DNSAddress2": "",
        "IPv4IPAddress": "192.168.1.1",
        "host_name": "hostname",
        "SubnetMask": "255.255.255.0",
        "DHCPServerStatus": 1,
        "StartIPAddress": "192.168.x.x",
        "EndIPAddress": "192.168.x.x",
        "DHCPLeaseTime": 24
    },
    "id": "12"
}
```

I think the wired ethernet adapter name can be controlled in Windows I believe, it's not something directly being influenced by the router I don't think. It's perhaps just the first label Windows assigned to the network profile. Windows 10 Settings menu is absolute rubbish these days, but I think it will be possible to change it, if not potentially going through the registry as a last resort.

**Comment by bit-user** [comment]
Did anybody have success in accessing the HH40 or reversing the encryption to upload a meaningful config file?

Thanks!

**Comment by LMQTLV** [comment]
Hello,
I am also interested for a connecting way to the HH40 (or more exactly HH40V for Europe), as none of the above methods are OK with actual firmware :(

**Comment by GangGames** [comment]
any idea how to remove the cell-id lock

**Comment by Ahxoohx** [comment]
> @dandare1980 For some reason there are firewall rules to stop direct access to the modem at 192.168.225.1 from the LAN side. It is kind of pointless as you can bypass this by logging into the router first then to the modem as you've found out or alternatively using the IPv4 address configured by the modem which is usually 19.x.x.x. (visible in the web interface).
> 
> It may not support direct access for USB or require some debug mode enabling to do it. There is adb, but it seems to lock up when you try and access TCP 5555.

It's support USB access,  but you must do everything in right way:

1. Connect cat4 cable to router and pc;
2. Press & hold Power & WPS;
3. Connect USB cable to pc (still holding buttons) ;
4. Release buttons when wifi led become active;

You will get an external drive, in which stored backup and firmware if there was update attempt.

And now you can access through ssh. Use address 192.168.225. 1  login with root and pass oelinux123


p.s.  If device bricked during update,  run: /etc/init.d/fota_progress start. It may need run that few times.

**Reply by etzisim** [reply]
after  /etc/init.d/fota_progress start my router only bootloops :(
dmesg doenst show it as any usb device... i need a way to get it back to fastboot mode so i can flash the modem firmware again

**Comment by c2vi** [comment]
Hi,
I have a hh40 from A1 Telekom Austria (software version:  HH40_ER_02.00_29) and i'm trying to gain ssh access for some custom dns.

I have now ssh access to the modem (root@mdm9607). Does anyone know how to now gain access to the router part?
How does modem and router work together? or communicate?

My configure.bin also seems to be encrypted, but in a different way than the one from @LoboHacks. (it only contains big/small letters, numbers, "/" and "+" with "==" at the end. Binwalk shows nothing as it is a normal textfile?)

In case anyone does not yet know: in the modem filesystem in /extern/backup_conf/ lays an unencrypted version of a zip (the one with the /etc/config/dropbear file) that should also be in the configure.bin

**Comment by killerbees19** [comment]
> it only contains big/small letters, numbers, "/" and "+" with "==" at the end.

[Base64](https://en.wikipedia.org/wiki/Base64)?

**Reply by c2vi** [reply]
Yup, Base64. After piping through base64 -d i got an openssl encrypted file, just like @LoboHacks . He actually did the same, but i just overlooked that.

**Comment by andyboeh** [comment]
In case someone is still interested in this platform: I found the encryption key in `core_app` using Ghidra and I'm able to decrypt the backup. If all goes well, I can enable SSH access to the router. If this works, I will add OpenWrt support for this Router. The shell access is definitely required because the U-Boot load address needs to be changed. A modern OpenWrt kernel doesn't fit into the 1,5MB kernel partition of the OEM firmware.

I will post the key and instructions a bit later - I have to figure out how to get this going with the current version of OpenSSL. For now, it only works with OpenSSL 1.0.

**Reply by LMQTLV** [reply]
Yes please, I'm still interested :)

**Reply by andyboeh** [reply]
To decrypt a backup file: `openssl aes-128-cbc -d -k oacgnahsnauyilil -base64 -in configure.bin -out configure.out -md md5`

This also works with the current OpenSSL release.
I'm currently preparing a Python script that modifies a backup file with SSH access enabled - it already works locally, I have to polish it a bit before posting it.

**Reply by andyboeh** [reply]
I published the Python script to modify a backup file to my Github: https://gist.github.com/andyboeh/d295c80a57d62379b926640762f3d5dd

Create a backup, then run `enable_sshd.py configure.bin` and upload the resulting `configure.bin_mod.bin` back to your router. Reboot and SSH is enabled. Default credentials: root:oelinux123

**Reply by andyboeh** [reply]
Here is my documentation on the backup file format and how to install OpenWrt: https://www.aboehler.at/doku/doku.php/projects:hh40v
I will create a PR for OpenWrt soon and put the relevant information into the OpenWrt wiki.

**Reply by killerbees19** [reply]
@andyboeh Nice work! 👍

Maybe it's time to find my device 😁

**Reply by LMQTLV** [reply]
@andyboeh thank you very much 👍 
I will dig up my old box and try that during my next holidays ...

**Reply by c2vi** [reply]
@andyboeh Great work!! 
How do you find UART Pins that are so hidden? Any Tips? I did look at the board of my hh40 and thought there isn't any UART.

**Reply by andyboeh** [reply]
With an oscilloscope it's pretty easy to identify Tx since the boot loader almost always prints some info on the UART. But to be honest, I was about to give up before I tried these pads since I expected three holes or pads in a row for Rx, Tx and GND. I tried nevertheless because someone in the OpenWrt forum mentioned a few years ago that there was a UART pad, but didn't give any information where it is.

**Reply by killerbees19** [reply]
> The LTE modem is an RNDIS device and works out-of-the-box if it was properly set up for the stock firmware. Limited configuration can be achieved using AT commands over USB (/dev/ttyUSB1). 

@andyboeh So there's currently no user-friendly method to change the LTE config (i.e. APN) after flashing OpenWrt? 🤓

But it's still possible to extend the predefined PLMN configs or change the custom profile(s), isn't it? (As mentioned [here](https://github.com/froonix/HH40V/wiki/RNDIS-config).)

My original and nearly unused device from 2018 (A1/SIMfonie) still has the old firmware, never updated. Backups are **not** encrypted and I can use SSH with `root:root`. I've saved all 7 MTD partitions and I'm ready to install OpenWrt…

* Software version: HH40_ER_02.00_23
* Modem version: HH40_ER_02.00_23

**Reply by andyboeh** [reply]
Sorry, I put the device in a storage box back in April and I don't remember the details - toying with the DreiNeo series was too tempting (and they have proper QMI modems inside). Please try it yourself and report back.

**Reply by killerbees19** [reply]
@andyboeh Works like a charm. Thanks for your work! 👍

The LTE modem works [OOTB](https://github.com/froonix/HH40V/wiki/OpenWrt-(23.05.0)) and it's still possible to change options at the SQLite database via SSH at the modem.

_Example: I accidentally forgot to re-enable the auto-connect flag at the original firmware, the modem no longer connects. I changed [one row](https://github.com/froonix/HH40V/wiki/RNDIS-config#auto-connect) at the database and rebooted the device, now it's working again. I haven't setup APN (etc.) and it works OOTB with B.free and bob SIM cards._

**Reply by killerbees19** [reply]
@andyboeh Just one more thing: Have you tested a reboot inside OpenWrt at any point?

It seems like it hangs while rebooting. Sadly I've no UART access at the moment to check the console output.

(But it boots again after disconnecting the power.)

_**EDIT**: But it's possible to reboot the 4G modem (`ssh root@192.168.225.1 reboot`), which results in a full device reboot. ^^_

**Reply by andyboeh** [reply]
It is very likely that I missed some restart GPIO - after all, it was a quick port and I don't use it regularly.

**Comment by Caresz93** [comment]
Hi There!
I got this Alcatel HH40V router. I think it's bricked, because only the power led lights up, not booting. I tried the WPS+power button method, where the router search firmware file on TFTP. I used a stock firmware, and some openwrt bin files, but nothing changes. Maybe someone have a working firmware bin file? Thanks for help!

**Reply by andyboeh** [reply]
...does it request an image via TFTP? Where did you get the firmware file from?
You cannot flash OpenWrt this way, the resulting image won't boot due to missing U-Boot settings.
You can find the TFTP recovery instructions at https://www.aboehler.at/doku/doku.php/projects:hh40v#restoring_stock

If it doesn't work, you will have to attach serial console to see what's going on.

**Reply by Caresz93** [reply]
Yes, i used Tftpd64, and the exact same instruction that you sent. I set the fix ip, then turn on router with wps+power buttons. The Tftpd64 pops up, and tell that sending had started. When reaches 100% the router restarts and do nothing. Still only power led lights.
I got the stock fw from [this link](https://alcatelfirmware.com/alcatel-onetouch-hh40v)

How can I get the serial console log? If it's the "UART" method, I have no equipment for this :(

**Comment by hellopanda17** [comment]
DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             OpenSSL encryption, salted, salt: 0xF1F1E04970265B60
If DECIMAL is 0 and HEXADECIMAL is 0x0 that means that file is empty
